### ISSUE TITLE
ASP.NET core 6.0 Kestrel ASPNETCORE_URLS env variable not working when set from Process.StartInfo.EnvironmentVariables

### ISSUE THREAD

**Issue by anulok99** [dotnet/aspnetcore#47563]
``I've already read this issue/thread https://github.com/dotnet/aspnetcore/issues/15011 however my issue seems different.

I have a main process, that will create child processes (the codebase of parent and child are different, but each child codebase is identical) where each child process needs to listen on a different kestrel port. 

I can open 4 CMD's in Windows Terminal and in each CMD window I can set the env variable for each window:
1st CMD terminal: set ASPNETCORE_URLS=http://0.0.0.0:15000;https://0.0.0.0:15001
2nd CMD terminal: set ASPNETCORE_URLS=http://0.0.0.0:15002;https://0.0.0.0:15003
3rd CMD terminal:  set ASPNETCORE_URLS=http://0.0.0.0:15004;https://0.0.0.0:15005
4th CMD terminal: set ASPNETCORE_URLS=http://0.0.0.0:15006;https://0.0.0.0:15007

Here is a portion of code from Program.cs for these child processes to prove I am correctly getting env variable set:

              ` foreach (DictionaryEntry de in Environment.GetEnvironmentVariables())
                {
                    Serilog.Log.Information("Environment variable {0} = {1} ", de.Key, de.Value);
                }

                var host = CreateHostBuilder(args).Build();

                var config = host.Services.GetRequiredService<IConfiguration>();

                foreach (var c in config.AsEnumerable())
                {
                    Console.WriteLine("configuration: " + c.Key + " = " + c.Value);
                }

                host.Run();`

where that will output: **[13:31:13 INF] Environment variable ASPNETCORE_URLS = http://0.0.0.0:15000;https://0.0.0.0:15001**
and:                              **configuration: ASPNETCORE_URLS = http://0.0.0.0:15000;https://0.0.0.0:15001**
for the first CMD terminal and all of that works fine for the other expected ports on the second, third, and fourth CMD windows, I can end up with all 4 running on different ports. 

**PROBLEM:** Now when I go to launch these same child processes from a parent process (where the parent process is launched from Visual Studio) it still wants to use default port of 5000 for all the children processes, here is the code that launches these children from a loop inside the Configure method of parent process:

               `foreach (var childProcess in allNeedeChildProcesses)
                {
                    Process jvmDriver = new Process();
                    jvmDriver.StartInfo.UseShellExecute = false;
                    jvmDriver.StartInfo.CreateNoWindow = false;
                    jvmDriver.StartInfo.WindowStyle = ProcessWindowStyle.Normal;
                    jvmDriver.StartInfo.EnvironmentVariables["ASPNETCORE_ENVIRONMENT"] = "Production";
                    int packagerHttpsPortNumber = Convert.ToInt32(childProcess.PortNumber);
                    jvmDriver.StartInfo.EnvironmentVariables["ASPNETCORE_URLS"] = "http://*:" + (packagerHttpsPortNumber - 1) + ";https://*:" + packagerHttpsPortNumber;
                    jvmDriver.StartInfo.FileName = @"..\JVMDriver\bin\Debug\net6.0\JVMDriver.exe";
                    jvmDriver.Start();
                    
                }

`
The output in console from parent process, which is now showing output from the first child process it just launched again shows:
**[13:49:37 INF] Environment variable ASPNETCORE_URLS = http://*:15000;https://*:15001**
and **configuration: ASPNETCORE_URLS = http://*:15000;https://*:15001** 
but then you see this output which shows it's still going to listen on port 5000:

`[13:49:37 INF] JVMDriver Hosting environment is Production
warn: Microsoft.AspNetCore.Server.Kestrel[0]
      Overriding address(es) 'http://*:15000, https://*:15001'. Binding to endpoints defined via IConfiguration and/or UseKestrel() instead.

info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://0.0.0.0:5000
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: https://0.0.0.0:5001
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
      Content root path: C:\Work\DevOps\CAPS\Master\Source\PUPSCore`

so logically you now expect when the 2nd child process is launched it throws exception saying port already in use:

[13:52:55 INF] JVMDriver dev: False prod: True
[13:52:55 INF] JVMDriver Hosting environment is Production
warn: Microsoft.AspNetCore.Server.Kestrel[0]
      Overriding address(es) 'http://*:15002, https://*:15003'. Binding to endpoints defined via IConfiguration and/or UseKestrel() instead.
[13:52:55 ERR] JVMDriver startup exception: System.IO.IOException: Failed to bind to address http://0.0.0.0:5000: address already in use.

The child process host builder code is simply:

         `public static IHostBuilder CreateHostBuilder(string[] args) =>
            Host.CreateDefaultBuilder(args)
                .ConfigureWebHostDefaults(webBuilder =>
                {
                    webBuilder.UseStartup<Startup>();
                });`

**Comment by Tratcher** [comment]
> Overriding address(es) 'http://:15000, https://:15001'. Binding to endpoints defined via IConfiguration and/or UseKestrel() instead.

Your environment variables are flowing, but they're conflicting with a different data source. There are a few different inputs that can configure Kestrel. ASPNETCORE_URLS is the simplest. When you dump the config in the child process are there any key that start with "Kestrel:Endpoints"?
https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-7.0#configureiconfiguration

**Comment by anulok99** [comment]
Yes, in fact I do, and I do see the 5000 and 5001 ports specified in those lines, however, it's still not clear why the env variable ASPNETCORE_URLS is not overriding them. I'm also not sure why those 2 lines are coming from. In the child process, I do have an appsettings.json and an appsettings.Development.json but neither of those is specifying 5000/5001. I do not have an appsettings.Production.json but if you don't have that json file and you run in production it's just going to use appsettings.json anyway. In the link you sent, it says **"No configuration
Kestrel listens on http://localhost:5000."** However, I have a configuration for kestrel in appsettings.json, but again it doesnt specify 5000/5001. 

configuration: Kestrel:Endpoints:Https:Url = https://0.0.0.0:5001
configuration: Kestrel:Endpoints:Https:Protocols = Http2
configuration: Kestrel:Endpoints:Https:Certificate =
configuration: Kestrel:Endpoints:Http =
configuration: Kestrel:Endpoints:Http:Url = http://0.0.0.0:5000
configuration: Kestrel:Endpoints:Http:Protocols = Http1
configuration: HTTPS_PORT = 5001
configuration: ENVIRONMENT = Production
configuration: ASPNETCORE_URLS = http://*:15000;https://*:15001

**Comment by anulok99** [comment]
Update: I just noticed that the above appsettings.json I pasted says "caps.pfx" for the path, and the one outputted in console window says "passcore.pfx", well guess what, the "passcore.pfx" is in fact defined in the parent process appsettings, the "caps.pfx" is specified in child appsettings. So when the parent process starts the child process, and the child process is the one that spits out the configuration, it's spitting out the parent's configuration, not it's own.

**Comment by Tratcher** [comment]
> Content root path: C:\Work\DevOps\CAPS\Master\Source\PUPSCore`

Is this the parent process dir? That's where it will load the appsettings from. Try setting the process's WorkingDirectory when you launch the child process.

**Comment by anulok99** [comment]
Yes, changing the working directory did fix the issue of it not using the child's appsettings, so now it is correctly using child appsettings. Before I ran I changed child's appsettings to use port 19000/19001, and it does in fact use those, but now we are back to the original problem in that I am setting ASPNETCORE_URLS as an environment variable from Process.StartInfo.EnvironmentVariables and it's not overrwrting the 19000/19001 in child process, as you can see the ASPNETCORE_URLS is correct in the configuration it spits out:

configuration: Kestrel =
configuration: Kestrel:Endpoints =
configuration: Kestrel:Endpoints:Https =
configuration: Kestrel:Endpoints:Https:Url = https://0.0.0.0:19001
configuration: Kestrel:Endpoints:Https:Protocols = Http2
configuration: Kestrel:Endpoints:Https:Certificate =
configuration: Kestrel:Endpoints:Https:Certificate:Path = caps.pfx
configuration: Kestrel:Endpoints:Https:Certificate:Password = *********
configuration: Kestrel:Endpoints:Https:Certificate:AllowInvalid = True
configuration: Kestrel:Endpoints:Http =
configuration: Kestrel:Endpoints:Http:Url = http://0.0.0.0:19000
configuration: Kestrel:Endpoints:Http:Protocols = Http1
**configuration: ASPNETCORE_URLS = http://*:15000;https://*:15001**
configuration: ASPNETCORE_HTTPS_PORT = 5001
configuration: ASPNETCORE_HOSTINGSTARTUPASSEMBLIES = Microsoft.AspNetCore.Watch.BrowserRefresh;Microsoft.WebTools.BrowserLink.Net
configuration: ASPNETCORE_ENVIRONMENT = Production
configuration: applicationName = JVMDriver
configuration: APPDATA = C:\Users\eschmitt\AppData\Roaming
configuration: ALLUSERSPROFILE = C:\ProgramData
configuration: AllowedHosts = *
configuration:  =
[14:15:47 INF] JVMDriver dev: False prod: True
[14:15:47 INF] JVMDriver Hosting environment is Production
warn: Microsoft.AspNetCore.Server.Kestrel[0]
      Overriding address(es) 'http://*:15000, https://*:15001'. Binding to endpoints defined via IConfiguration and/or UseKestrel() instead.
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://0.0.0.0:19000
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: https://0.0.0.0:19001
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
      Content root path: C:\Work\DevOps\CAPS\Master\Source\JVMDriver\bin\Debug\net6.0

**Comment by Tratcher** [comment]
Kestrel:Endpoints:Https:Url is higher priority than ASPNETCORE_URLS, that's by design. That's why you get the `Overriding address(es)` warning. To override 19001, set "Kestrel:Endpoints:Https:Url=https://*:15001" as an environment variable.

**Comment by anulok99** [comment]
Finally, that worked. Here is my working child process launcher:

            `foreach (var childProcess in allNeedeChildProcesses)
                {
                    Process jvmDriver = new Process();
                    jvmDriver.StartInfo.UseShellExecute = false;
                    jvmDriver.StartInfo.CreateNoWindow = false;
                    jvmDriver.StartInfo.WindowStyle = ProcessWindowStyle.Normal;
                    jvmDriver.StartInfo.EnvironmentVariables["ASPNETCORE_ENVIRONMENT"] = "Production";
                    int packagerHttpsPortNumber = Convert.ToInt32(childProcess.PortNumber);
                    //jvmDriver.StartInfo.EnvironmentVariables["ASPNETCORE_URLS"] = "http://*:" + (packagerHttpsPortNumber - 1) + ";https://*:" + packagerHttpsPortNumber;
                    jvmDriver.StartInfo.EnvironmentVariables["Kestrel:Endpoints:Https:Url"] = "https://*:" + packagerHttpsPortNumber;
                    jvmDriver.StartInfo.EnvironmentVariables["Kestrel:Endpoints:Http:Url"] = "https://*:" + (packagerHttpsPortNumber - 1);
                    jvmDriver.StartInfo.WorkingDirectory = @"..\JVMDriver\bin\Debug\net6.0\";
                    jvmDriver.StartInfo.FileName = @"..\JVMDriver\bin\Debug\net6.0\JVMDriver.exe";
                    jvmDriver.Start();
                    
                }`

So in normal operation, say just 1 process, it reads from asppsettings, like below example and the runtime makes an env variable from the below into "Kestrel:Endpoints:Https:Url=https://localhost:15001"?

{
  "Kestrel": {
    "Endpoints": {
      "Http": {
        "Url": "http://localhost:5000"
      },
      "Https": {
        "Url": "https://localhost:5001"
      }
    }
  }
}

Looking at order of precedence: 
ASP.NET Core includes several providers for loading configuration from a variety of sources including:

JSON Files
XML Files
INI Files
Command-line arguments
Environment variables
In-memory .NET objects
Secret Manager storage
Encrypted in Azure Key Vault

If the env variable is constructed from appsettings.json, (assuming you come back and say yes to this question) then really which precedence is it falling under, appsettings.json or env variable in the above precedence list? One is coming from the other. Also, I've never seen in documentation, including the link you sent yesterday any literature on one environment variable, namely "Kestrel:Endpoints:Https:Url" overriding environment "ASPNETCORE_URLS".

**Comment by Tratcher** [comment]
It's not a very big note:

> These endpoints replace those defined in the top-level Urls configuration rather than adding to them.

https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-7.0#replace-the-default-certificate-from-configuration

> If the env variable is constructed from appsettings.json

The IConfiguration is built in the precedence order you described above. However, it only composes the different inputs, `Kestrel:Endpoints:Https:Url` and `ASPNETCORE_URLS` are still separate config entries that are checked later by the server. It's kestrel that's accessing both of these values from the config and prioritizing the more specific Kestrel:Endpoints config over the more general ASPNETCORE_URLS.
